### PR TITLE
feat(2658): introduce `Adapter`

### LINKED ISSUE #2658: Update `wizard` to programatically control the LLM adapter and model
Currently based on environment variables the GenAI library uses a particular Adapter and Model. We wish to change that and provide a type-safe way to integrate it.

**Example**
```rust
let model = Adapter::Groq::Llama3_8b_8192;
let wiz = Wizard::new(model);
```

**Technical Requirements**
- `Adapter` shouild be defined in `llm/adapter.rs`.
- Secrets should be read as `TAILCALL_SECRET` and sent to the LLM

### PR DESCRIPTION
**Pull request by ssddOnTop**
**Issue Reference(s):**  
Fixes #2658
/claim 2658

### CODE REVIEW

File: src/cli/llm/adapter.rs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+#[derive(Clone)]
+pub enum Adapter {
+    Groq(GroqModel),
+}
+
+#[derive(Clone)]
+pub enum GroqModel {
+    Llama38b8192,
+}
```
Comment (tusharmath) [inline src/cli/llm/adapter.rs:11]: Add All supported models and adapters that the library support.

Comment (ssddOnTop) [reply src/cli/llm/adapter.rs:11]: is there any list for that?

File: src/cli/generator/generator.rs [hunk 2]
```diff
@@ -166,7 +166,9 @@ impl Generator {
 
         if infer_type_names {
             let mut llm_gen = InferTypeName::default();
```
Comment (tusharmath) [inline src/cli/generator/generator.rs:168]:
```suggestion
            let secret = self.runtime.env.get("TAILCALL_SECRET");
            let mut llm_gen = InferTypeName::new(secret);
```

File: src/cli/llm/wizard.rs [hunk 3]
```diff
@@ -1,27 +1,29 @@
...
 
 use super::Result;
+use crate::cli::llm::model::Model;
 
 #[derive(Setters, Clone)]
 pub struct Wizard<Q, A> {
     client: Client,
-    // TODO: change model to enum
-    model: String,
+    model: Model,
```
Comment (tusharmath) [inline src/cli/llm/wizard.rs:12]: Should optionally hold the secret.

File: src/cli/llm/model.rs [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+        AdapterKind::from_model(self.0).unwrap()
+    }
+}
+
+impl OpenAI {
+    const GPT4O: Model = Model("gpt-4o");
+    const GPT4O_MINI: Model = Model("gpt-4o-mini");
+    const GPT4_TURBO: Model = Model("gpt-4-turbo");
+    const GPT4: Model = Model("gpt-4");
+    const GPT35_TURBO: Model = Model("gpt-3.5-turbo");
```
Comment (tusharmath) [inline src/cli/llm/model.rs:39]: don't need const any more here.

File: src/cli/llm/infer_type_name.rs [hunk 5]
```diff
@@ -1,13 +1,14 @@
...
 
 use super::{Error, Result, Wizard};
+use crate::cli::llm::adapter::Adapter;
+use crate::cli::llm::model::Model;
 use crate::core::config::Config;
 
-const MODEL: &str = "llama3-8b-8192";
-
 #[derive(Default)]
 pub struct InferTypeName {}
```
Comment (tusharmath) [inline src/cli/llm/infer_type_name.rs:13]: Can hold an instance of the wizard over here with the secret.

**Review by tusharmath — APPROVED** [review]

**Pull request merged**